lib/DataLayoutAnalysis/Frontend/DLACreateInterProceduralTypes.cpp

```diff
@@ -105,51 +102,44 @@ bool TSBuilder::createInterproceduralTypes(llvm::Module &M,
       auto [ArgNode, _] = getOrCreateLayoutType(&ArgVal);
       revng_assert(ArgNode);
 
-      model::QualifiedType ArgumentModelType;
+      model::UpcastableType ArgumentModelType;
       if (RFT) {
         const auto &ModelArgs = RFT->Arguments();
-        auto NumModelArguments = ModelArgs.size();
-        if (ArgIndex < NumModelArguments) {
-          auto ArgIt = std::next(ModelArgs.begin(), ArgIndex);
-          ArgumentModelType = ArgIt->Type();
-        } else {
-          ArgumentModelType = { RFT->StackArgumentsType(), {} };
-        }
+        if (ArgIndex < ModelArgs.size())
+          ArgumentModelType = std::next(ModelArgs.begin(), ArgIndex)->Type();
+        else
+          ArgumentModelType = RFT->StackArgumentsType();
       } else {
         revng_assert(CABIFT);
-        const auto &Args = CABIFT->Arguments();
-        ArgumentModelType = Args.at(ArgIndex).Type();
+        ArgumentModelType = CABIFT->Arguments().at(ArgIndex).Type();
       }
 
-      if (ArgumentModelType.UnqualifiedType().isValid()
-          and ArgumentModelType.isPointer()) {
-        const model::QualifiedType
-          Pointee = peelConstAndTypedefs(ArgumentModelType).stripPointer();
-
-        bool IsScalar = Pointee.isScalar();
-
-        auto MaybeSize = Pointee.size();
-        bool IsSized = MaybeSize.has_value();
-
-        namespace TDKind = model::TypeDefinitionKind;
-        bool IsFunction = Pointee.is(TDKind::RawFunctionDefinition)
-                          or Pointee.is(TDKind::CABIFunctionDefinition);
-
-        // If it is not scalar then it must be sized or a function type
-        revng_assert(IsScalar or IsSized or IsFunction);
-
-        if (not IsScalar) {
-          ArgNode->NonScalar = true;
-          if (IsSized)
-            ArgNode->Size = *MaybeSize;
-          else if (IsFunction)
-            ArgNode->Size = getPointerSize(Model.Architecture());
-        } else {
-          // Skip char, because they alias and propagate weird information.
-          if (IsSized and *MaybeSize > 1)
-            ArgNode->Size = *MaybeSize;
-          else if (IsFunction)
-            ArgNode->Size = getPointerSize(Model.Architecture());
+      if (model::PointerType *Pointer = ArgumentModelType->getPointer()) {
+        if (Pointer->PointeeType() && Pointer->PointeeType()->verify()) {
```

Comment:
:thinking: How is the pointee type even supposed to not verify?

Reply:
When it's broken? But yeah, I guess, that's not possible, if it's coming from a valid model - since it's always a piece of an RFT or a CFT - and we already verified those.

Reply:
Exactly

Reply:
Done.